### DISCUSSION TITLE
s-exp/inline comment syntax

### DISCUSSION THREAD

**Discussion by iacore** [janet-lang/janet#1272] [Ideas]
Scheme has [`#;(...)`](https://srfi.schemers.org/srfi-62/srfi-62.html) that allows a block of code to be commented easily.

Please share your idea about how such syntax would be in Janet.

`(comment ...)` is not the same, since it returns `nil`.

**Comment by primo-ppcg** [comment]
Something like this would have worked until recently:
```janet
(defmacro ignore [& body]
  ~;[])

(pp (ignore
      this is just a comment.)
    :ok)
```
```
:ok
```
However, this will throw an error as a top level expression:
```
(ignore
  this is just a comment.)
```
```
compile error: splice can only be used in function parameters and data constructors, it has no effect here
```

**Reply by iacore** [reply]
So the proposal is to allow top-level splice? Or empty macro as a special case.

I think we do need reader syntax for this case.

**Reply by primo-ppcg** [reply]
The error produced by an improper usage of a splice is meant to prevent accidental propagation of a splice form. I think it would be preferable to have an empty macro expand to nothing, rather than nil, although this won't be done.

**Reply by bakpakin** [reply]
The reason this doesn't generally work is that your comment would only work in certain places. For example:

```
(def (ignore a b c) x 10)
```

would never have worked. BTW, the posted example does actually work in the latest janet. It is the equivalent of `(pp ;[]  :ok)` which is well formed.

**Comment by pepe** [comment]
Just my two cents, but for this purpose, I use good old `#`. I understand that it is not the same as having a particular form, but I use this technique sparingly and mostly lean on git.

**Reply by iacore** [reply]
`#` in Janet is line comment
this discussion is about s-exp/inline comment

**Reply by pepe** [reply]
Yep, as I wrote, I understand that it is not the same, just that I am using it for this purpose.

**Reply by pepe** [reply]
Anyway, I see little value in this whole discussion.

**Comment by sogaiu** [comment]
I am not opposed to the idea of block comments that don't evaluate to nil, but I'm not enthusiastic about new syntax.

I can imagine using such a feature and can relate, but so far have survived without it.

I don't know what costs might be involved in carrying out something like this though -- both within Janet itself but also to adapting existing tooling.

AFAIU neither Clojure nor Emacs Lisp have something like this [2] and Scheme does [1], so some Lisp-using folks seem to get by with less.  It might be that spelling out (in some detail) the kinds of situations where this idea would be useful could be helpful in getting support.

I think sometimes you don't know what you're missing until you see good uses for it, though of course not always (^^;

Regarding new syntax, [this issue](https://github.com/janet-lang/janet/issues/1184) contains a fairly recent discussion.  It might be helpful background in understanding various folks' views.

_Update_: added footnote [2] after [reducing own confusion](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6860832).

---
[1] No clue about Common Lisp.

[2] Actually, Clojure has `#_` which AFAICT is just like SRFI-62's `#;` (note that neither require delimiters afterwards).

**Reply by CosmicToast** [reply]
Honestly if I was to do this, I would make `comment` a special form that evaluates to nothing and call it a day.

**Reply by sogaiu** [reply]
Please don't change the existing behavior of `comment`.

If there were code that was like this:

```janet
(my-fn a b (comment c))
```

If `my-fn` expects 3 arguments and can handle a third one being `nil`, that would break.

Apart from unforeseen consequences, I think a shorter name [1] would be nicer for the proposed construct anyway.

---

[1] `ignore` is shorter, though even that seems long for something one might want to use to quickly disable something without utilizing special editor support.

**Reply by pepe** [reply]
Just to clear myself, I have seen in my time with Janet many people come with: in the language X this or that works this or that way, change it in Janet! At the beginning, even I felt the same from time to time. Now, with better understanding of the whole process and the time investment in my code I am very cautious of any of these rants. 

I am sorry, but the special form for this seems to me as an overkill, especially changing the functionality in a lot of existing code. As has been reminded not that long ago: nobody cares about breaking an old code, as long as it is not his own. Reader macros idea is also not a very bright idea for Janet in my opinion.

And as always: everything Janet is a fair game.

**Reply by sogaiu** [reply]
Not sure what rant you are referring to.

I think there is just a request for a certain kind of functionality and some people are just exploring how it might be done and others are expressing caution (myself included).

I don't know what the costs might be for such a feature, but if they were low enough, I would be happy to see such functionality.

I suspect that folks who edit their source and are sending expressions via a REPL connection to a Janet process are likely to appreciate this type of functionality [1] more than if using a more edit-save-run type of flow, so it may be that how much value one feels may be affected by such factors.

---

[1] Tweaking code here and there for exploration seems to be lower cost this way (at the cost of state-related issues messing things up).

**Reply by iacore** [reply]
Thanks for the link. I don't ever use the `|1` syntax. `(fn [])` is good enough for me.

```
(my-fn a b (comment c))
```

This example is janky. I wonder if anyone write code like this intentionally.

**Reply by pepe** [reply]
> Not sure what rant you are referring to.

I used the wrong word, I guess. I meant this answer: I would make comment a special form that evaluates to nothing and call it a day.

My apologies rainy day by the sea 😅

**Reply by iacore** [reply]
[I tried it, and made `(comment)` a special form.](https://github.com/iacore/janet/tree/comment-special) It's not quite working for top-level code.

It functions like a comment at top-level, but evaluates to `@[]` (visible in REPL).

I'm sure that can be fixed.

**Reply by CosmicToast** [reply]
> My apologies rainy day by the sea 😅

Yeah, that's not really a rant. A rant refers to a tirade, a lengthy, wild, passionate speech, which can often be boisterous, nonsensical, and raving.
Indeed, I meant it as a brief comment of little value besides offering one potential solution. :)

**Reply by primo-ppcg** [reply]
> Note that this can't be a special form easily either

It's essentially the same as the macro.

**Reply by bakpakin** [reply]
The macro/special form does not work and cannot work in a reasonable way. All other special forms and macros would need to understand how to handle comments since they would be retained by the parser. The correct and simple way to do this is to extend the parser. Janet also will not support arbitrary reader macros.

**Comment by bakpakin** [comment]
Possibly this is a good use (if somewhat underwhelming) of our last free special character in the janet syntax, `\`.

Something like

```janet
(def \(This is just a fancy comment) a 10)
```

Would discard the comment during reading. Backwards compatible and simple.

**Reply by bakpakin** [reply]
Note that this can't be a special form easily either, it needs to be something built into the parser that will prevent forms from ever reaching the compiler and macro expander. Unlike other prefixes like `;`, `,`, and `~`, the backslash would just pop the form off of the parse stack rather than replace it with a 2-element tuple.

The only caveat is that it might be better to save backslash some more, err, valuable syntax

**Reply by primo-ppcg** [reply]
A big advantage of syntax like this (aside from not evaluating to `nil`) is that a multi-line form can be commented out at a single location, rather than having to add a matching parenthesis at the end.

I'm not sold on `\`, though. I experimented briefly with [multi-byte reader macros](https://github.com/primo-ppcg/janet/commit/a27247fe3d9388da32785af16ae21801f467cee7#diff-d011fa541ea7b4129b467d4bbfd75111facaf4e2e61b7e4c35d327b50bc27453R193), and it works well enough. It wouldn't need to be a single character, which makes many other options available.
```
'#(is this a comment?)

,#(is this a comment?)

;#(is this a comment?)

|#(is this a comment?)

~#(is this a comment?)
```
Any of these could be added easily. I would say that `;#(...)` looks the nicest.

**Reply by iacore** [reply]
It will work like `|`, but tells the parser to ignore the next expression.

`;#(...)` won't work, because

```
; #(this is comment)
[splice content here]
```

Anything with `#` will change the current behavior (which is to ignore anything behind it on the same line)

**Reply by primo-ppcg** [reply]
Theoretically. I believe it would be exceptionally rare to separate a splice reader macro from its target by so much as a single space, let alone a comment.

That said, the second character wouldn't need to be `#`, although I do think this makes it abundantly clear that it is a comment.

**Reply by sogaiu** [reply]
Tossing out a couple of related ideas:

* As mentioned by @iacore (below / elsewhere), other languages sort-of dedicate a character to a kind of extension mechanism.  Clojure also does this with `#`.  Perhaps using `\` to introduce an extension "sequence" (multibyte reader macro as mentioned above) is pretty valuable.  Isn't it possible some other feature / syntax idea will pop up down the line?

* Clojure also has a "discard" feature `#_` that is in some sense more general than the multi-line comment and is more convenient because one does not need delimiters surrounding the item to be "thrown out":

   ```
   (+ 1 2 #_ 3)
   # =>
   3
   ```

   But you can put a list (and ofc other dual-delimiter constructs) after the `#_` as well:

   ```
   (+ 1 2 #_ (- 1 1))
   # =>
   3
   ```

   This is because it "discards" the next thing "read".

---

As a side note, a nice feature of a construct like `#_`, is that it allows one to quickly "toggle" between two argument forms in a call:

```
(my-fn one-alt #_ another-alt)
```

```
(my-fn #_ one-alt another-alt)
```

via transposition of a form (assuming one's editor supports this).

**Reply by primo-ppcg** [reply]
> Clojure also has a "discard" feature `#_`

I guess that's what we're talking about. As far as using backslash for this, I think that "underwhelming" is an accurate assessment. I think that something like `;#` would be a much better choice, despite being - theoretically - not 100% backwards compatible.

**Reply by sogaiu** [reply]
I didn't spell it out above, but I think the two ideas mentioned can be combined -- so `\` can be used as an extension introducer and one of those uses could be for discarding.  That's not underwhelming to me because it leaves a path open for future additions.

I would also say it's not the same as what was proposed above because IIUC that idea requires delimiters on both sides.  In my experience, it's very nice ergnomically not to have to bother with the closing delimiter.

I don't know how important it is for this case, but it does happen to be backward compatible as well.

In any case, my main motivation above was to put some things on the table so that they might participate in the ultimate decision -- whether accepted or not, I would rather they have been part of the discussion (this is something I've been waiting to talk about for a couple of years (^^; ).

**Reply by sogaiu** [reply]
(Unfortunately, some of these comments that are "deeper" may not be seen by others -- one must notice that there is a link to "click" to see them...not a great characteristic for a system that's supposed to be promoting discussions IMO.)

**Reply by sogaiu** [reply]
I looked at some of the items that [@iacore linked to](https://srfi.schemers.org/?keywords=reader-syntax) [1] and came across [SRFI 62: S-expression comments](https://srfi.schemers.org/srfi-62/srfi-62.html).  The construct described above -- `#_` -- seems pretty similar to it.  The SRFI one uses `#;` instead.

The abstract from the SRFI is:

> This SRFI proposes a simple extension to Scheme's lexical syntax that allows individual S-expressions to be made into comments, ignored by the reader. This contrasts with the standard Lisp semicolon comments, which make the reader ignore the remainder of the line, and the slightly less common block comments, as [SRFI 30](https://srfi.schemers.org/srfi-30/srfi-30.html) defines: both of these mechanisms comment out slices of text, not S-expressions.

The rationale is:

> Line and block comments are useful for embedding textual commentary in programs, but they are not conducive to commenting out code easily in an absence of extensive editor support for removing selected text that composes S-expressions while retaining them in the text itself, or subsequently removing the comments and re-introducing the S-expressions themselves. 

[SRFI 62](https://srfi.schemers.org/srfi-62/srfi-62.html) is a pretty short read (and I've quoted a lot it already) ;)

I'll note that the OP has this text:

> it allows a block of code to be commented easily.

So I read that concern as having to do more with commenting out code and not necessarily text.

---

[1] In [this comment](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6844735).

**Reply by iacore** [reply]
Yes, `#;` is mainly used to comment out code in scheme. You can use it to write block comment as well.

**Reply by sogaiu** [reply]
My impression so far is that this kind of construct -- `#;` (SRFI 62) / `#_` (Clojure) -- would be nicer, at least from a user perspective.  No detailed thoughts about implementation consequences.

For reference, IIRC (memory a bit fuzzy) supporting `#_` in tree-sitter-clojure involved redoing the grammar.  So I know one way it can be accomplished -- using that way for tree-sitter-janet-simple will mean breaking changes (which might be worth it).

**Reply by sogaiu** [reply]
LOL, sorry, I've been misreading.  I see that the OP is mentioning SRFI 62 (which I presume was what was there originally).

I think part of my mistake was that I didn't look / read closely enough (definitely my bad - sorry!) and assumed wrongly that the parens and dots implied another construct.

@iacore Please consider editing the OP to use `#;` only -- so without the parens and dots.

**Comment by iacore** [comment]
Slightly related: Scheme's myriad reader syntax.

In Scheme, `#` has special meaning. `#.(Pi)` is implementation-defined data. `#()` is vector. `#;()` is comment. `#<p></p>` is XML. [More here](https://srfi.schemers.org/?keywords=reader-syntax)

We don't have such a convention in Janet yet, so it's difficult to add new syntax.

P.S. Common Lisp has `#.( )`

**Comment by sogaiu** [comment]
So to clarify, IIUC -- SRFI 62's `#;` and Clojure's `#_`:

* "operate on" / "affect" the "next" expression [1] -- not arbitrary text
* neither have parens or other dual-delimiters as part of the construct, though one CAN follow the construct with a paren-delimited expression

This is different from typical multi-line comment constructs (e.g. `/*` + `*/` in C-ish languages) that "comment out" text which can include strings that aren't "source", and it's different from `(comment ...)` (which does have an opening and closing delimiter, sort of -- and the `...` can have multiple expressions).

Some examples (`#_` is a stand-in, it wouldn't be a good choice [2]):

```
(+ 1 #_ 2 8)  # just the 2 "disappears"

# parses to
(+ 1 8)
```

```
#_ first item goes away

# parses to
item goes away
```

```
(+ 1 #_ (all of this stuff disappears but each item between parens is treated as a symbol first))

# parses to
(+ 1)
```

```
{:alice "restaurant"
 #_ #_ :bob "bunker"
 :charles "choco-factory"}

# parses to
{:alice "restaurant"
 :charles "choco-factory"}
```

A demo of [primo-ppcg's comment](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6843947):

> A big advantage of syntax like this (aside from not evaluating to nil) is that a multi-line form can be commented out at a single location, rather than having to add a matching parenthesis at the end.

```
(defn my-fn
  [x]
  (print "hi")
  {:fox [:f (* x 2)]
   :elephant [:e math/pi]}
  #_{:ant [:a x]
     :bee [:b 2]}) # note: we don't have to figure out which closing delimiters here with #_.
                   #       using (comment ...) need to carefully place the closing paren
                   #       because of the presence of ] and }
# parses to
(defn my-fn
  [x]
  (print "hi")
  {:fox [:f (* x 2)]
   :elephant [:e math/pi]})
```

(I'm writing this here because stuff can become obscure and hidden in GH discussions (at lower levels) and [I was confused about what was being proposed in the OP](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6860832) (perhaps it was just me (^^; ).)

---

[1] Note that "stacking" can happen.

For example, for:

```
(+ 1 #_ #_ 2 3 6)
```

the inner `#_` applies to `2` and the outer `#_` applies to `3`.

```
(+ 1 #_ #_ 2 3 6)
     |  |  | |
     |  ---- |
     |       |
     ---------
```

The net result is:

```
(+ 1 6)
```

[2] (_Note_: this footnote was added hours after the initial series of edits -- sorry, primo-ppcg (^^; )

As implied by others ("underwhelming"), I think using `\` alone (single character) just for this feature is not worth it.

If `\` is to be involved, I think it would be better to use [a mutli-byte sequence](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6843947) (similar to what primo-ppcg suggested in concept) where `\` is the first character.  (I think this arrangement would be backward compatible whereas it's unclear to me whether choosing some other leading character would be.)

Using `\` + single character seems similar in flavor to what iacore mentioned [here](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6844735) about the use of `#` in Scheme (e.g. `#()` is vector, `#;` discards, etc.).  This would potentially allow other sequences beginning with `\` in the future.

**Reply by sogaiu** [reply]
I worked on this a bit and did get something that appears to work (at least in limited testing).  I did this in a way that uses more than one character -- conceptually like what primo-ppcg suggested (though with `\#`).

I also tried to find and update code that might be affected (e.g. PEGs that process Janet source, tree-sitter grammar, spork/fmt, tests in janet, janet-lang.org docs, etc.).  Likely I didn't find all relevant things but it was an interesting exercise in getting some sense of what kinds of consequences a change in syntax might have in associated libraries, programs, docs, and such.

I came across an interesting phrase in the last few days -- ["programmer convenience feature"](https://youtu.be/zlWqpRTgXBg?t=320).  I think one can make a case for the feature in this discussion to be counted as one such.

I'm not sure how often such things are worth the costs (e.g. adding code, maintaining code, learning about the code if not yet familiar, impact on other code outside the main codebase, etc.).  Individually they might seem ok, but over time they might end up interacting poorly with more essential features and/or hinder/hurt/prevent some change that might have been very beneficial.

(Granted, there might be features that are both convenient and actually offer other things, but in this case, perhaps this feature is primarily "programmer convenience".)

On a related note, one of the things I've been doing in some Lisps and Lisp-likes is to prepend a single quote in front of some forms I wish to "comment out".  This doesn't work everywhere, but it does work in a number of situations, e.g.

```janet
# top-level form can be commented out
'(defn my-fn
  [x]
  (+ x 8))
```

or:

```janet
(defn fun-fun
  [a b]
  (def c 1)
  # some forms can be commented even if not at top-level
  '(pp "hi there")
  (+ a b))
```

**Reply by primo-ppcg** [reply]
I also spent some time thinking about this after the discussion and independently concluded that `\#` would likely be the best choice. The idea that I had was that `\` would start a different type of reader macro ("extended reader macro" or similar) that could allow any token to follow.

While this would utilize the last unassigned character, it would offer some flexibility, as the number of reader macros possible would be essentially limitless.

**Reply by sogaiu** [reply]
> The idea that I had was that \ would start a different type of reader macro ("extended reader macro" or similar) that could allow any token to follow.

It sounds conceptually similar to what iacore mentioned [here](https://github.com/janet-lang/janet/discussions/1272#discussioncomment-6844735), and I went for something like that in the [attempt](https://github.com/sogaiu/janet/commit/a590841b8998b9039396d7d0f0c45ea2b86d75b7).

Does that seem close to what you had in mind?

---

Before using Janet, I was of the opinion that I'd likely want to always (if possible) have some way to keep extending a programming language if necessary.  A supporting thought here is something like "who knows what need or idea might arise at some point?".

I'm not so sure I buy this line of thinking any more -- at least I'm not sure it doesn't make sense to start over at some point from time-to-time.  It seems likely to me that  potentially infinite extensibility may lead to needless complexity and awkward constructs.  I think those kinds of things tend to make languages harder to learn, reason about, and maintain, and that's a pretty undesirable thing as far as I'm concerned.

**Reply by pepe** [reply]
> I'm not so sure I buy this line of thinking any more -- at least I'm not sure it doesn't make sense to start over at some point from time-to-time. It seems likely to me that potentially infinite extensibility may lead to needless complexity and awkward constructs. I think those kinds of things tend to make languages harder to learn, reason about, and maintain, and that's a pretty undesirable thing as far as I'm concerned.

I feel the same more and more. 🙏

**Reply by primo-ppcg** [reply]
> Does that seem close to what you had in mind?

Almost exactly that.

> It seems likely to me that potentially infinite extensibility may lead to needless complexity and awkward constructs.

Not needing to have the language reference open constantly is the mark of good design. This is more true of Janet than most languages I've used.

Essentially limitless, yes, but how many more reader macros could there really be anyway? Perhaps two or three? My point was that by using `\` in this way, the door hasn't been closed forever. You could still add a `\&` or `\@` or whatever else might make sense in the future.

**Reply by sogaiu** [reply]
> My point was that by using \ in this way, the door hasn't been closed forever. You could still add a \& or \@ or whatever else might make sense in the future.

Sure - as I understand it, that's what at least 3 of us (you, iacore, and yours truly) seem to be concluding / saying in this discussion and I tried to reflect that ["open-endedness"](https://github.com/sogaiu/janet/commit/a590841b8998b9039396d7d0f0c45ea2b86d75b7#diff-d011fa541ea7b4129b467d4bbfd75111facaf4e2e61b7e4c35d327b50bc27453R602-R615) in the little coding experiment.

> Essentially limitless, yes, but how many more reader macros could there really be anyway? Perhaps two or three?

You may be right.  In some sense, I think zero would be the nicest, but perhaps if some compelling thing comes along I might change my mind :)

For the record, I looked a bit into this for Scheme and Clojure...

In Scheme's case it seems there were a bit more than two or three proposed.  iacore posted a [link](https://srfi.schemers.org/?keywords=reader-syntax) earlier that points at page that lists more than 10.  It's possible that none of those things actually made it into any of the RnRS standards -- but some folks were interested enough to write stuff up.  So with enough of the right people I guess there is sufficient imagination?

I'm not intimately familiar with the history of how things got added to Clojure, but at least currently, one can see what gets "dispatched" via `#` [here](https://github.com/clojure/clojure/blob/6975553804b0f8da9e196e6fb97838ea4e153564/src/jvm/clojure/lang/LispReader.java#L109-L120).

I think some time after 1.0-ish, at least 4 things got added:

* `#^` - for metadata (this was actually a change in the "official" notation IIUC)
* `##` - for symbolic values
* `#?` - for reader conditionals
* `#:` - for namespaced maps

That's 4 things that are handled by the dispatch reader macro-ish construct, but those were not the only additions to the language.  I think tagged literals also got added after 1.0.  These are just the things that I happen to be familiar with though.  Quite possibly I missed some other things.

I think for both Scheme and Clojure there is kind of an expectation of extensibility.
